src/m2n/PointToPointCommunication.cpp

```diff
@@ -496,6 +566,73 @@ void PointToPointCommunication::requestConnection(std::string const &acceptorNam
   _isConnected = true;
 }
 
+void PointToPointCommunication::requestPreConnection(std::string const &acceptorName,
+                                                     std::string const &requesterName)
+{
+  TRACE(acceptorName, requesterName);
+  CHECK(not isConnected(), "Already connected!");
+  CHECK(utils::MasterSlave::isMaster() || utils::MasterSlave::isSlave(),
+        "You can only use a point-to-point communication between two participants which both use a master. "
+        << "Please use distribution-type gather-scatter instead.");
+
+  if (utils::MasterSlave::isMaster()) {
+    // Establish connection between participants' master processes.
+    auto c = _communicationFactory->newCommunication();
+    c->requestConnection(acceptorName, requesterName, 0, 1);
+    
+  } else {
+    assertion(utils::MasterSlave::isSlave());
+  }
+
+  std::vector<int> localConnectedRanks = _mesh->getConnectedRanks();
+
+  if (localConnectedRanks.empty()) {
+    _isConnected = true;
+    return;
+  }
+
+#ifdef SuperMUC_WORK
+  try {
+    auto addressDirectory = _communicationFactory->addressDirectory();
+
+    utils::Parallel::synchronizeProcesses();
+  } catch (...) {
+  }
+#endif  
+
+  std::vector<com::PtrRequest> requests;
+  requests.reserve(localConnectedRanks.size());
+  _connectionDataVector.reserve(localConnectedRanks.size());
+
+  std::set<int> acceptingRanks;
+  for (auto &i : localConnectedRanks)
```

Comment:
Is there any reason to use `auto &` instead of `int`? In my opinion this makes it only harder to read.